File: core/BasicCalculatorII/BasicCalculatorII.java
package com.dbc;

import java.util.Stack;

public class BasicCalculatorII {
    public int calculate(String s) {
        Stack<Integer> stack = new Stack<>();
        int num = 0;
        char preOp = '+';
        for (int i = 0; i < s.length(); i++) {
            if (Character.isDigit(s.charAt(i))) {
                num = num * 10 + (s.charAt(i) - '0');
            }
            if (i == s.length() - 1 || s.charAt(i) == '-' || s.charAt(i) == '+' || s.charAt(i) == '*' || s.charAt(i) == '/') {
                if (preOp == '+') stack.add(num);
                else if (preOp == '-') stack.add(-num);
                else if (preOp == '*') stack.add(num * stack.pop());
                else stack.add(stack.pop() / num);
                num = 0;
                preOp = s.charAt(i);
            }
        }
        int res = 0;
        while (!stack.isEmpty()) res += stack.pop();
        return res;
    }
}

File: core/BasicCalculatorII/BasicCalculatorII.py
def calculate(s: str) -> int:
    stack, num, length, pre_op = [], 0, len(s), "+"
    for i in range(length):
        if s[i].isdigit():
            num = num * 10 + int(s[i])
        if i == length - 1 or s[i] in "+-*/":
            if pre_op == "+":
                stack.append(num)
            elif pre_op == "-":
                stack.append(-num)
            elif pre_op == "*":
                stack.append(stack.pop() * num)
            else:
                stack.append(int(stack.pop() / num))
            pre_op = s[i]
            num = 0

    return sum(stack)




if __name__ == '__main__':
    print(calculate("0-2147483647"))

